File: src/main/java/org/lanternpowered/server/service/permission/UserCollection.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission;

import static com.google.common.base.Preconditions.checkArgument;

import org.lanternpowered.server.game.Lantern;
import org.lanternpowered.server.service.permission.base.LanternSubject;
import org.lanternpowered.server.service.permission.base.LanternSubjectCollection;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.profile.GameProfile;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;

import java.util.Collection;
import java.util.UUID;

import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * User collection keeping track of opped users.
 */
public class UserCollection extends LanternSubjectCollection {

    UserCollection(LanternPermissionService service) {
        super(PermissionService.SUBJECTS_USER, service);
    }

    @Nullable
    private UUID parseUUID(String identifier) {
        try {
            return UUID.fromString(identifier);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    @Override
    public LanternSubject get(String identifier) {
        final UUID uuid = parseUUID(identifier);
        checkArgument(uuid != null, "Provided identifier must be a uuid, was %s", identifier);
        final GameProfile profile;
        try {
            profile = Sponge.getServer().getGameProfileManager().get(uuid, true).get();
        } catch (Exception e) {
            throw new IllegalArgumentException("Failed to lookup game profile for " + uuid, e);
        }
        return new UserSubject(profile, this);
    }

    @Override
    public boolean isRegistered(String identifier) {
        final UUID uuid = parseUUID(identifier);
        return uuid != null && Lantern.getGame().getOpsConfig().getEntryByUUID(uuid).isPresent();
    }

    @Override
    @SuppressWarnings({"unchecked", "rawtypes"})
    public Collection<Subject> getLoadedSubjects() {
        return (Collection) Sponge.getServer().getOnlinePlayers();
    }

    public LanternPermissionService getService() {
        return this.service;
    }
}

File: src/main/kotlin/org/lanternpowered/server/service/permission/UserCollection.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission

import org.lanternpowered.api.Server
import org.lanternpowered.server.game.Lantern
import org.spongepowered.api.service.permission.PermissionService
import org.spongepowered.api.service.permission.Subject
import java.util.UUID

/**
 * User collection keeping track of opped users.
 */
class UserCollection internal constructor(
        val service: LanternPermissionService
) : LanternSubjectCollection(PermissionService.SUBJECTS_USER, service) {

    private fun parseUniqueId(identifier: String): UUID? {
        return try {
            UUID.fromString(identifier)
        } catch (e: IllegalArgumentException) {
            null
        }
    }

    override fun get(identifier: String): LanternSubject {
        val uuid = parseUniqueId(identifier) ?: error("Provided identifier must be a uuid, was $identifier")
        val profile = try {
            Server.gameProfileManager[uuid, true].get()
        } catch (e: Exception) {
            throw IllegalArgumentException("Failed to lookup game profile for $uuid", e)
        }
        return UserSubject(profile, this)
    }

    override fun isRegistered(identifier: String): Boolean {
        val uuid = parseUniqueId(identifier)
        return uuid != null && Lantern.getGame().opsConfig.getEntryByUUID(uuid).isPresent
    }

    override fun getLoadedSubjects(): Collection<Subject> = Server.onlinePlayers
}
